include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h

```diff
@@ -508,8 +509,76 @@ using __repacked_tuple_t = typename __repacked_tuple<T>::type;
 template <typename _ContainerOrIterable>
 using __value_t = typename __internal::__memobj_traits<_ContainerOrIterable>::value_type;
 
-//A contract for future class: <sycl::event or other event, a value or sycl::buffers...>
-//Impl details: inheretance (private) instead of aggregation for enabling the empty base optimization.
+template <typename _T>
+struct __accessor
+{
+    using __accessor_t = sycl::accessor<_T, 1, sycl::access::mode::read_write, __dpl_sycl::__target_device,
+                                        sycl::access::placeholder::false_t>;
+    __accessor_t m_acc;
+    _T* m_ptr;
+    bool m_usm;
+
+    auto
+    get_pointer() const //should be cached within a kernel
+    {
+        return m_usm ? m_ptr : &m_acc[0];
+    }
+};
+
+template <typename _ExecutionPolicy, typename _T>
+struct __storage
+{
+  private:
+    using __sycl_buffer_t = sycl::buffer<_T, 1>;
+    ::std::shared_ptr<__sycl_buffer_t> m_sycl_buf;
+    ::std::shared_ptr<_T> m_usm_buf;
+    bool m_usm;
+
+  public:
+    __storage(_ExecutionPolicy& __exec, bool __usm, ::std::size_t __n) : m_usm(__usm)
+    {
+        if (m_usm)
+        {
+            m_usm_buf = std::shared_ptr<_T>(
+                __internal::__sycl_usm_alloc<_ExecutionPolicy, _T, sycl::usm::alloc::host>{__exec}(__n),
+                __internal::__sycl_usm_free<_ExecutionPolicy, _T>{__exec});
+        }
+        else
+            m_sycl_buf = ::std::make_shared<__sycl_buffer_t>(__sycl_buffer_t(__n));
+    }
+
+    auto
+    get_acc(sycl::handler& __cgh)
+    {
+        __accessor<_T> acc;
+        if (m_usm)
+        {
+            acc.m_usm = true;
```

Comment:
We can move this duplicated line outside of condition: ```acc.m_usm = m_usm;```

Reply:
Done.